### PR TITLE
chore(iast): ast Visitor as singleton

### PR DESCRIPTION
**Pull request by gnufede**
IAST: Performance improvement: Reuse AST Visitor object.

## Checklist

- [x] Change(s) are motivated and described in the PR description
- [x] Testing strategy is described if automated tests are not included in the PR
- [x] Risks are described (performance impact, potential for breakage, maintainability)
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Description motivates each change
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] Release note makes sense to a user of the library
- [x] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/appsec/_iast/_ast/ast_patching.py [hunk 1]
```diff
@@ -19,6 +19,9 @@
 from .visitor import AstVisitor
 
 
+VISITOR = AstVisitor()
```
Comment (gnufede) [inline ddtrace/appsec/_iast/_ast/ast_patching.py:22]:
```suggestion
_VISITOR = AstVisitor()
```

File: ddtrace/appsec/_iast/_ast/ast_patching.py [hunk 2]
```diff
@@ -99,13 +102,10 @@ def visit_ast(
 ) -> Optional[str]:
     parsed_ast = ast.parse(source_text, module_path)
 
-    visitor = AstVisitor(
-        filename=module_path,
-        module_name=module_name,
-    )
-    modified_ast = visitor.visit(parsed_ast)
+    VISITOR.update_location(filename=module_path, module_name=module_name)
+    modified_ast = VISITOR.visit(parsed_ast)
```
Comment (gnufede) [inline ddtrace/appsec/_iast/_ast/ast_patching.py:106]:
```suggestion
    _VISITOR.update_location(filename=module_path, module_name=module_name)
    modified_ast = _VISITOR.visit(parsed_ast)
```

File: ddtrace/appsec/_iast/_ast/ast_patching.py [hunk 3]
```diff
@@ -99,13 +102,10 @@ def visit_ast(
...
 
-    if not visitor.ast_modified:
+    if not VISITOR.ast_modified:
```
Comment (gnufede) [inline ddtrace/appsec/_iast/_ast/ast_patching.py:108]:
```suggestion
    if not _VISITOR.ast_modified:
```

File: ddtrace/appsec/_iast/_ast/visitor.py [hunk 4]
```diff
@@ -126,14 +128,14 @@ def _mark_avoid_convert_recursively(node):
             "super",
         },
     },
-}
+}  # type: Dict
```
Comment (avara1986) [inline ddtrace/appsec/_iast/_ast/visitor.py:131]: `# type` is deprecated in favor of `_ASPECTS_SPEC: Dict = {`

File: ddtrace/appsec/_iast/_ast/visitor.py [hunk 5]
```diff
@@ -173,6 +171,15 @@ def __init__(
...
         self._taint_sink_replace_disabled = _ASPECTS_SPEC["taint_sinks"]["disabled"]
 
+        self.update_location(filename, module_name)
+
+    def update_location(self, filename: str = "", module_name: str = ""):
+        self.filename = filename
+        self.module_name = module_name
+        self.ast_modified = False
+
+        excluded_from_patching = _ASPECTS_SPEC["excluded_from_patching"]  # type: Dict[str, Dict[str, Tuple[str]]]
```
Comment (avara1986) [inline ddtrace/appsec/_iast/_ast/visitor.py:181]: dito

**Review by juanjux — APPROVED** [review]

**Pull request merged**